docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond
+  directly to MachineSets and MachineDeployments that are marked for autoscaling.
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters. One feature that is missing from this integration is
+the ability to scale down to, and up from, a MachineSet or MachineDeployment with zero replicas.
+
+This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define
+the specific resource requirements for each Infrastructure Machine Temalate they create. These
+requirements are then utilized by the cluster autoscaler to make decisions about the number
+of nodes to add during scale up operations.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated cloud providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and cloud providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Provide a mechanism for users to override the defaults for any given MachineSet or MachineDeployment.
+- Create an optional API contract that defines the minimum needed information for scaling from and to zero replicas.
+
+### Non-Goals/Future Work
+
+- Create an API contract that cloud providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Create additional CRDs or controllers
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
+this information to create a prediction about how many nodes should be created when scaling. In
+most sitatuations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
+
+An optional field is proposed for users to add to their Infrastructure Machine Templates which will
```

Comment:
If that field is similar in all infrastructure providers, is there something blocking us from having it directly in MachineSet spec? I understand that infrastructure template can be reused but won't having it in MS spec make it more obvious that it is scaled from zero? WDYT?

Reply:
in general, i tend to agree with you. it would be nice to have this information only on the MachineSet or MachineDeployment resources, but one problem here is that we want this information to be more closely related to the individual infrastructure providers since they will need to supply this data.

when we discussed this issue previously, one issue that arose was the idea that each cloud provider should be able to opt-in to the scale from zero behavior depending on how their cloud implementation is structured. by placing this information in the Machine[Set|Deployment] we are moving it further from the provider implementations and closer to the users.

the idea that we had settled on during previous discussions was that the individual providers could populate this field on their infrastructure templates (depending on the machine sizes available in that cloud), and then it would be reconciled into the Machine[Set|Deployment] status. in this manner we expose the information on the Machine[Set|Deployment] but also make it clear that it is fixed information at that level.

the hope here is that we will reduce the need for users to directly control these values and allow them to be part of the infrastructure providers implementation details.